File: src/prism/inline.cpp
#include "prism/stages.hpp"
#include "prism/regex.hpp"

#include <cctype>
#include <filesystem>
#include <map>
#include <optional>
#include <string>
#include <string_view>
#include <tuple>
#include <unordered_set>
#include <utility>
#include <vector>

namespace prism {
namespace {

const std::unordered_set<std::string> kKw = {
    "if", "for", "while", "switch", "return", "sizeof", "typeof",
    "else", "do", "case", "default", "_Generic", "break", "continue",
    "goto", "struct", "union", "enum",
};

const char* kScalarTypes =
    "int|unsigned(?:\\s+int)?|long|short|char|"
    "uint32_t|int32_t|size_t|bool|_Bool";

using Index = std::map<std::pair<std::string, std::string>, const FunctionInfo*>;

std::string strip(std::string s) {
    std::size_t a = 0, b = s.size();
    while (a < b && std::isspace(static_cast<unsigned char>(s[a]))) ++a;
    while (b > a && std::isspace(static_cast<unsigned char>(s[b - 1]))) --b;
    return s.substr(a, b - a);
}

std::string lstrip_copy(std::string text) {
    std::size_t i = 0;
    while (i < text.size() && std::isspace(static_cast<unsigned char>(text[i]))) ++i;
    return text.substr(i);
}

std::string squeeze_ws(std::string_view typ) {
    std::string out;
    bool sp = false;
    for (char c : typ) {
        if (std::isspace(static_cast<unsigned char>(c))) {
            if (!sp && !out.empty()) out.push_back(' ');
            sp = true;
        } else {
            out.push_back(c);
            sp = false;
        }
    }
    return out;
}

std::string basename_of(const std::string& file) {
    return std::filesystem::path(file).filename().string();
}

bool is_ident(std::string_view name) {
    static Regex re("[A-Za-z_]\\w*");
    auto m = re.search_match(name, 0);
    return m && !m->spans.empty() && m->spans[0].first == 0 &&
           static_cast<std::size_t>(m->spans[0].second) == name.size();
}

// Calls the BMC encoder models (bmc_encoder.inc model_call): a callee that
// only calls these can still be inlined. __VERIFIER_nondet_* is matched by prefix.
const std::unordered_set<std::string> kModelledCalls = {
    "abs", "labs", "llabs", "__builtin_expect", "rand", "__VERIFIER_assume",
    "assume_abort_if_not", "abort", "exit", "_Exit", "quick_exit", "reach_error",
    "__assert_fail", "__VERIFIER_error", "printLine", "printWLine", "printIntLine", "printShortLine",
    "printLongLine", "printLongLongLine", "printSizeTLine", "printHexCharLine", "printUnsignedLine", "printHexUnsignedCharLine"};

bool has_calls(std::string_view body) {
    static Regex re("\\b([A-Za-z_]\\w*)\\s*\\(");
    for (auto& m : re.finditer(body)) {
        auto n = m.group(1);
        if (kKw.contains(n) || kModelledCalls.contains(n) || n.starts_with("__VERIFIER_nondet_")) continue;
        return true;
    }
    return false;
}

bool returns_value(const FunctionInfo& callee) {
    std::string rt = strip(callee.return_type.empty() ? "int" : callee.return_type);
    return rt != "void" && !rt.ends_with(" void");
}

// Integer type names the BMC encoder declares (bmc_encoder.inc CDECL_TYPE).
const char* kDeclType =
    "(?:(?:unsigned|signed)\\s+(?:long\\s+long|long|short|char|int)(?:\\s+int)?|"
    "long\\s+long(?:\\s+int)?|long(?:\\s+int)?|short(?:\\s+int)?|"
    "unsigned|signed|int|char|_Bool|bool|u?int(?:8|16|32|64)_t|"
    "size_t|ssize_t|ptrdiff_t|u?intptr_t|u?intmax_t)";

std::string regex_sub(const Regex& re, const std::string& src, const std::string& repl);

// The callee's full return type, from its signature (return_type keeps only
// the last word: `unsigned char` would read as `char`). nullopt: not a type
// the encoder can declare, so the callee is not inlined.
std::optional<std::string> callee_ret_type(const FunctionInfo& callee) {
    auto sig = callee.signature;
    auto at = sig.find(callee.name + "(");
    if (at == std::string::npos) at = sig.find(callee.name);
    if (at == std::string::npos) return std::nullopt;
    auto head = sig.substr(0, at);
    static Regex attrs("\\[\\[[^\\]]*\\]\\]|\\b(?:static|inline|__inline__|__inline|extern|constexpr)\\b");
    head = squeeze_ws(strip(regex_sub(attrs, head, " ")));
    if (head == "void") return head;
    static Regex full(std::string("^") + kDeclType + "$");
    auto m = full.search_match(head, 0);
    if (!m) return std::nullopt;
    return head;
}

bool inlineable_callee(const FunctionInfo& callee) {
    if (!callee.is_static) return false;
    if (callee.kind != "SCALAR" && callee.kind != "VOID") return false;
    if (has_calls(callee.body)) return false;
    if (returns_value(callee) && !callee_ret_type(callee)) return false;
    return true;
}

const FunctionInfo* lookup(const Index& index, const std::string& file, const std::string& name) {
    auto it = index.find({basename_of(file), name});
    return it == index.end() ? nullptr : it->second;
}

std::vector<std::string> split_args(std::string args) {
    args = strip(std::move(args));
    if (args.empty()) return {};
    std::vector<std::string> out;
    int depth = 0;
    std::size_t start = 0;
    for (std::size_t i = 0; i < args.size(); ++i) {
        char ch = args[i];
        if (ch == '(') ++depth;
        else if (ch == ')') --depth;
        else if (ch == ',' && depth == 0) {
            out.push_back(strip(args.substr(start, i - start)));
            start = i + 1;
        }
    }
    out.push_back(strip(args.substr(start)));
    return out;
}

std::optional<std::tuple<std::string, std::string, std::size_t>> parse_call(const std::string& text,
                                                                          std::size_t start) {
    static Regex ident("[A-Za-z_]\\w*");
    if (start > text.size()) return std::nullopt;
    auto m = ident.search_match(std::string_view(text).substr(start), 0);
    if (!m || m->spans.empty() || m->spans[0].first != 0) return std::nullopt;
    std::string name = m->text;
    if (kKw.contains(name) || !is_ident(name)) return std::nullopt;
    std::size_t pos = start + name.size();
    while (pos < text.size() && (text[pos] == ' ' || text[pos] == '\t' || text[pos] == '\n')) ++pos;
    if (pos >= text.size() || text[pos] != '(') return std::nullopt;
    int depth = 0;
    for (std::size_t i = pos; i < text.size(); ++i) {
        char ch = text[i];
        if (ch == '(') ++depth;
        else if (ch == ')') {
            --depth;
            if (depth == 0) {
                return std::make_tuple(name, text.substr(pos + 1, i - (pos + 1)), i + 1);
            }
        }
    }
    return std::nullopt;
}

std::optional<std::pair<std::string, std::string>> match_balanced(const std::string& text, char open_ch,
                                                                 char close_ch) {
    if (text.empty() || text.front() != open_ch) return std::nullopt;
    int depth = 0;
    for (std::size_t i = 0; i < text.size(); ++i) {
        if (text[i] == open_ch) ++depth;
        else if (text[i] == close_ch) {
            --depth;
            if (depth == 0) {
                return std::make_pair(text.substr(1, i - 1), text.substr(i + 1));
            }
        }
    }
    return std::nullopt;
}

std::pair<std::string, std::string> take_stmt(const std::string& text) {
    int depth = 0;
    for (std::size_t i = 0; i < text.size(); ++i) {
        char ch = text[i];
        if (ch == '(') ++depth;
        else if (ch == ')') --depth;
        else if (ch == '{' && depth == 0) break;
        else if (ch == ';' && depth == 0) {
            return {text.substr(0, i + 1), text.substr(i + 1)};
        }
    }
    return {text, {}};
}

std::string regex_sub(const Regex& re, const std::string& src, const std::string& repl) {
    std::string out;
    std::size_t i = 0;
    for (auto& m : re.finditer(src)) {
        if (m.spans.empty()) continue;
        auto a = static_cast<std::size_t>(m.spans[0].first < 0 ? 0 : m.spans[0].first);
        auto b = static_cast<std::size_t>(m.spans[0].second < 0 ? 0 : m.spans[0].second);
        if (a < i) continue;
        out.append(src, i, a - i);
        for (std::size_t k = 0; k < repl.size(); ++k) {
            if ((repl[k] == '$' || repl[k] == '\\') && k + 1 < repl.size() &&
                std::isdigit(static_cast<unsigned char>(repl[k + 1]))) {
                out += m.group(static_cast<std::size_t>(repl[k + 1] - '0'));
                ++k;
            } else {
                out += repl[k];
            }
        }
        i = b;
    }
    out.append(src, i, std::string::npos);
    return out;
}

std::string rename_params(const std::string& body, const std::map<std::string, std::string>& rename) {
    static Regex ident("[A-Za-z_]\\w*");
    std::string out;
    std::size_t i = 0;
    for (auto& m : ident.finditer(body)) {
        if (m.spans.empty()) continue;
        auto a = static_cast<std::size_t>(m.spans[0].first < 0 ? 0 : m.spans[0].first);
        auto b = static_cast<std::size_t>(m.spans[0].second < 0 ? 0 : m.spans[0].second);
        if (a < i) continue;
        out.append(body, i, a - i);
        auto it = rename.find(m.text);
        out += (it != rename.end()) ? it->second : m.text;
        i = b;
    }
    out.append(body, i, std::string::npos);
    return out;
}

// A `return` must leave the inlined body: every return becomes
// `{ ret = X; break; }` inside `do { ... } while (0)`. A callee with its own
// loop or switch (where that break would bind to the wrong statement) is
// inlined only when its single return is its last top-level statement.
std::optional<std::string> adapt_callee_body(std::string body, const std::map<std::string, std::string>& rename,
                                             const std::optional<std::string>& ret_var) {
    body = strip(rename_params(body, rename));
    static Regex ret_any("\\breturn\\b");
    static Regex ret_val("\\breturn\\s+([^;]+);");
    static Regex ret_bare("\\breturn\\s*;");
    static Regex loops("\\b(?:for|while|do|switch)\\b");
    auto rets = ret_any.finditer(body);
    if (rets.empty()) return body;
    bool single_final = false;
    if (rets.size() == 1) {
        auto at = static_cast<std::size_t>(rets[0].spans[0].first);
        int depth = 0;
        for (std::size_t i = 0; i < at; ++i) {
            if (body[i] == '{') ++depth;
            else if (body[i] == '}') --depth;
        }
        auto semi = body.find(';', at);
        single_final = depth == 0 && semi != std::string::npos && strip(body.substr(semi + 1)).empty();
    }
    if (single_final) {
        if (ret_var) body = regex_sub(ret_val, body, *ret_var + " = $1;");
        return strip(regex_sub(ret_bare, body, ""));
    }
    if (!loops.finditer(body).empty()) return std::nullopt;
    if (ret_var) body = regex_sub(ret_val, body, "{ " + *ret_var + " = $1; break; }");
    body = regex_sub(ret_bare, body, "break;");
    return "do {\n" + body + "\n} while (0);";
}

// Locals the callee declares, renamed with the call-site prefix so they do
// not collide with the caller's names.
void add_callee_locals(const std::string& body, const std::string& prefix,
                       std::map<std::string, std::string>& rename) {
    static Regex decl(std::string("\\b") + kDeclType + "\\s+([A-Za-z_]\\w*)\\s*(?=[=;,\\[)])");
    static const std::unordered_set<std::string> type_words = {
        "int", "unsigned", "signed", "long", "short", "char", "_Bool", "bool"};
    for (auto& m : decl.finditer(body)) {
        auto name = m.group(1);
        if (type_words.contains(name) || rename.count(name)) continue;
        rename[name] = prefix + "_l_" + name;
    }
}

std::string param_type(const std::string& typ) {
    auto t = squeeze_ws(typ.empty() ? "int" : typ);
    return t.empty() ? "int" : t;
}

std::optional<std::string> build_inline_block(const FunctionInfo& callee,
                                              const std::vector<std::string>& args,
                                              const std::string& prefix,
                                              const std::string& tail) {
    if (args.size() != callee.params.size()) return std::nullopt;
    std::vector<std::string> decls;
    std::map<std::string, std::string> rename;
    for (std::size_t i = 0; i < callee.params.size(); ++i) {
        const auto& [typ, pname] = callee.params[i];
        std::string temp = prefix + "_i" + std::to_string(i);
        rename[pname] = temp;
        decls.push_back(param_type(typ) + " " + temp + " = " + args[i] + ";");
    }
    std::optional<std::string> ret_var;
    if (returns_value(callee)) {
        auto rt = callee_ret_type(callee);
        if (!rt) return std::nullopt;
        ret_var = prefix + "_ret";
        decls.push_back(*rt + " " + *ret_var + ";");
    }
    add_callee_locals(callee.body, prefix, rename);
    auto adapted = adapt_callee_body(callee.body, rename, ret_var);
    if (!adapted) return std::nullopt;
    std::vector<std::string> parts = std::move(decls);
    if (!adapted->empty()) parts.push_back(std::move(*adapted));
    if (!tail.empty()) {
        std::string t = tail;
        while (!t.empty() && t.back() == ';') t.pop_back();
        parts.push_back(t + ";");
    }
    std::string inner;
    for (std::size_t i = 0; i < parts.size(); ++i) {
        if (i) inner += "\n";
        inner += parts[i];
    }
    return "{\n" + inner + "\n}";
}

std::string try_inline_stmt(const std::string& stmt, const FunctionInfo& fn, const Index& index,
                            int site) {
    const std::string& raw = stmt;
    std::string s = strip(stmt);
    if (s.empty() || s.back() != ';') return raw;

    std::string prefix = "_h" + std::to_string(site);

    static Regex ret_re("return\\s+");
    if (auto m = ret_re.search_match(s, 0); m && !m->spans.empty() && m->spans[0].first == 0) {
        auto end = static_cast<std::size_t>(m->spans[0].second);
        auto call = parse_call(s, end);
        if (!call) return raw;
        auto [name, args_src, call_end] = *call;
        auto rest_src = s.substr(end);
        std::size_t k = 0;
        while (k < rest_src.size() && std::isspace(static_cast<unsigned char>(rest_src[k]))) ++k;
        if (!rest_src.substr(k).starts_with(name)) return raw;
        auto* callee = lookup(index, fn.file, name);
        if (!callee || callee->name == fn.name || !inlineable_callee(*callee)) return raw;
        if (strip(s.substr(call_end)) != ";") return raw;
        auto args = split_args(args_src);
        std::string tail = "return " + std::string(returns_value(*callee) ? prefix + "_ret" : "0");
        auto block = build_inline_block(*callee, args, prefix, tail);
        return block ? *block : raw;
    }

    static Regex assign_re("([A-Za-z_]\\w*)\\s*=\\s*");
    if (auto m = assign_re.search_match(s, 0); m && !m->spans.empty() && m->spans[0].first == 0) {
        std::string lhs = m->group(1);
        auto end = static_cast<std::size_t>(m->spans[0].second);
        auto call = parse_call(s, end);
        if (!call) return raw;
        auto [name, args_src, call_end] = *call;
        auto* callee = lookup(index, fn.file, name);
        if (!callee || callee->name == fn.name || !inlineable_callee(*callee)) return raw;
        if (!returns_value(*callee)) return raw;
        if (strip(s.substr(call_end)) != ";") return raw;
        auto args = split_args(args_src);
        auto block = build_inline_block(*callee, args, prefix, lhs + " = " + prefix + "_ret");
        return block ? *block : raw;
    }

    static Regex decl_re(std::string("(?P<typ>") + kScalarTypes +
                         ")\\s+(?P<name>[A-Za-z_]\\w*)\\s*=\\s*");
    if (auto m = decl_re.search_match(s, 0); m && !m->spans.empty() && m->spans[0].first == 0) {
        std::string lhs = m->named("name");
        auto end = static_cast<std::size_t>(m->spans[0].second);
        auto call = parse_call(s, end);
        if (!call) return raw;
        auto [name, args_src, call_end] = *call;
        auto* callee = lookup(index, fn.file, name);
        if (!callee || callee->name == fn.name || !inlineable_callee(*callee)) return raw;
        if (!returns_value(*callee)) return raw;
        if (strip(s.substr(call_end)) != ";") return raw;
        auto args = split_args(args_src);
        auto block = build_inline_block(*callee, args, prefix, lhs + " = " + prefix + "_ret");
        // The declaration stays in the caller's scope; the block assigns it.
        return block ? m->named("typ") + " " + lhs + ";\n" + *block : raw;
    }

    auto call = parse_call(s, 0);
    if (!call) return raw;
    auto [name, args_src, call_end] = *call;
    if (strip(s.substr(call_end)) != ";") return raw;
    auto* callee = lookup(index, fn.file, name);
    if (!callee || callee->name == fn.name || !inlineable_callee(*callee)) return raw;
    if (returns_value(*callee)) return raw;
    auto args = split_args(args_src);
    auto block = build_inline_block(*callee, args, prefix, "");
    return block ? *block : raw;
}

std::string transform_body(std::string text, const FunctionInfo& fn, const Index& index, int& counter) {
    std::string out;
    while (!text.empty()) {
        text = lstrip_copy(std::move(text));
        if (text.empty()) break;
        if (text.front() == '{') {
            auto matched = match_balanced(text, '{', '}');
            if (!matched) {
                out += text;
                break;
            }
            out += '{';
            out += transform_body(matched->first, fn, index, counter);
            out += '}';
            text = matched->second;
            continue;
        }
        auto [stmt, rest] = take_stmt(text);
        text = rest;
        if (strip(stmt).empty()) continue;
        auto new_stmt = try_inline_stmt(stmt, fn, index, counter);
        if (new_stmt != stmt) ++counter;
        out += new_stmt;
    }
    return out;
}

FunctionInfo inline_function(FunctionInfo fn, const Index& index) {
    if (fn.kind == "POINTER") return fn;
    int counter = 0;
    auto new_body = transform_body(fn.body, fn, index, counter);
    if (counter == 0) return fn;
    fn.body = std::move(new_body);
    // The inlined body no longer maps offset for offset onto the source.
    fn.body_line = fn.body_col = 0;
    return fn;
}

}  // namespace

std::vector<FunctionInfo> inline_static(const std::vector<FunctionInfo>& functions) {
    Index index;
    for (auto& f : functions) index[{basename_of(f.file), f.name}] = &f;
    std::vector<FunctionInfo> out;
    out.reserve(functions.size());
    for (auto& fn : functions) out.push_back(inline_function(fn, index));
    return out;
}

}  // namespace prism

File: prism/inline.py
"""One-level inliner for static scalar callees in the same translation unit."""

from __future__ import annotations

from dataclasses import replace
from pathlib import Path
import re

from prism.models import FunctionInfo

_KW = {
    "if", "for", "while", "switch", "return", "sizeof", "typeof",
    "else", "do", "case", "default", "_Generic", "break", "continue",
    "goto", "struct", "union", "enum",
}

_SCALAR_TYPES = (
    r"int|unsigned(?:\s+int)?|long|short|char|"
    r"uint32_t|int32_t|size_t|bool|_Bool"
)


def _basename(file: str) -> str:
    return Path(file).name


def _is_ident(name: str) -> bool:
    return bool(re.fullmatch(r"[A-Za-z_]\w*", name))


# Calls the BMC encoder models (bmc._model_call): a callee that only calls
# these can still be inlined. __VERIFIER_nondet_* is matched by prefix.
_MODELLED_CALLS = {
    "abs", "labs", "llabs", "__builtin_expect", "rand", "__VERIFIER_assume",
    "assume_abort_if_not", "abort", "exit", "_Exit", "quick_exit", "reach_error",
    "__assert_fail", "__VERIFIER_error", "printLine", "printWLine", "printIntLine", "printShortLine",
    "printLongLine", "printLongLongLine", "printSizeTLine", "printHexCharLine", "printUnsignedLine", "printHexUnsignedCharLine",
}


def _has_calls(body: str) -> bool:
    for m in re.finditer(r"\b([A-Za-z_]\w*)\s*\(", body or ""):
        n = m.group(1)
        if n in _KW or n in _MODELLED_CALLS or n.startswith("__VERIFIER_nondet_"):
            continue
        return True
    return False


def _returns_value(callee: FunctionInfo) -> bool:
    rt = (callee.return_type or "int").strip()
    return rt != "void" and not rt.endswith(" void")


# Integer type names the BMC encoder declares (bmc._CDECL_TYPE).
_DECL_TYPE = (
    r"(?:(?:unsigned|signed)\s+(?:long\s+long|long|short|char|int)(?:\s+int)?|"
    r"long\s+long(?:\s+int)?|long(?:\s+int)?|short(?:\s+int)?|"
    r"unsigned|signed|int|char|_Bool|bool|u?int(?:8|16|32|64)_t|"
    r"size_t|ssize_t|ptrdiff_t|u?intptr_t|u?intmax_t)"
)


def _callee_ret_type(callee: FunctionInfo) -> str | None:
    """The callee's full return type, from its signature.

    `return_type` keeps only the last word (`unsigned char` would read as
    `char`). None: not a type the encoder can declare, so no inlining.
    """
    sig = callee.signature or ""
    at = sig.find(callee.name + "(")
    if at < 0:
        at = sig.find(callee.name)
    if at < 0:
        return None
    head = sig[:at]
    head = re.sub(r"\[\[[^\]]*\]\]|\b(?:static|inline|__inline__|__inline|extern|constexpr)\b",
                  " ", head)
    head = " ".join(head.split())
    if head == "void":
        return head
    if not re.fullmatch(_DECL_TYPE, head):
        return None
    return head


def _inlineable_callee(callee: FunctionInfo) -> bool:
    if not callee.static:
        return False
    if callee.kind not in ("SCALAR", "VOID"):
        return False
    if _has_calls(callee.body):
        return False
    if _returns_value(callee) and _callee_ret_type(callee) is None:
        return False
    return True


def _lookup(
    index: dict[tuple[str, str], FunctionInfo],
    file: str,
    name: str,
) -> FunctionInfo | None:
    return index.get((_basename(file), name))


def _split_args(args: str) -> list[str]:
    args = args.strip()
    if not args:
        return []
    out: list[str] = []
    depth = 0
    start = 0
    for i, ch in enumerate(args):
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
        elif ch == "," and depth == 0:
            out.append(args[start:i].strip())
            start = i + 1
    out.append(args[start:].strip())
    return out


def _parse_call(text: str, start: int) -> tuple[str, str, int] | None:
    m = re.match(r"[A-Za-z_]\w*", text[start:])
    if not m:
        return None
    name = m.group(0)
    if name in _KW or not _is_ident(name):
        return None
    pos = start + len(name)
    while pos < len(text) and text[pos] in " \t\n":
        pos += 1
    if pos >= len(text) or text[pos] != "(":
        return None
    depth = 0
    for i in range(pos, len(text)):
        ch = text[i]
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
            if depth == 0:
                return name, text[pos + 1 : i], i + 1
    return None


def _match_balanced(text: str, open_ch: str, close_ch: str) -> tuple[str, str] | None:
    if not text.startswith(open_ch):
        return None
    depth = 0
    for i, ch in enumerate(text):
        if ch == open_ch:
            depth += 1
        elif ch == close_ch:
            depth -= 1
            if depth == 0:
                return text[1:i], text[i + 1 :]
    return None


def _take_stmt(text: str) -> tuple[str, str]:
    depth = 0
    for i, ch in enumerate(text):
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
        elif ch == "{" and depth == 0:
            break
        elif ch == ";" and depth == 0:
            return text[: i + 1], text[i + 1 :]
    return text, ""


def _rename_params(body: str, rename: dict[str, str]) -> str:
    """One pass over the identifiers (the C++ engine's rename_params)."""
    return re.sub(r"[A-Za-z_]\w*", lambda m: rename.get(m.group(0), m.group(0)), body)


def _adapt_callee_body(body: str, rename: dict[str, str], ret_var: str | None) -> str | None:
    """A `return` must leave the inlined body.

    Every return becomes `{ ret = X; break; }` inside `do { ... } while (0)`.
    A callee with its own loop or switch (where that break would bind to the
    wrong statement) is inlined only when its single return is its last
    top-level statement; otherwise None (not inlined).
    """
    body = _rename_params(body, rename).strip()
    rets = list(re.finditer(r"\breturn\b", body))
    if not rets:
        return body
    single_final = False
    if len(rets) == 1:
        at = rets[0].start()
        depth = body[:at].count("{") - body[:at].count("}")
        semi = body.find(";", at)
        single_final = depth == 0 and semi >= 0 and not body[semi + 1:].strip()
    if single_final:
        if ret_var:
            body = re.sub(r"\breturn\s+([^;]+);", lambda m: f"{ret_var} = {m.group(1)};", body)
        return re.sub(r"\breturn\s*;", "", body).strip()
    if re.search(r"\b(?:for|while|do|switch)\b", body):
        return None
    if ret_var:
        body = re.sub(r"\breturn\s+([^;]+);", lambda m: f"{{ {ret_var} = {m.group(1)}; break; }}", body)
    body = re.sub(r"\breturn\s*;", "break;", body)
    return "do {\n" + body + "\n} while (0);"


def _add_callee_locals(body: str, prefix: str, rename: dict[str, str]) -> None:
    """Locals the callee declares, renamed with the call-site prefix."""
    type_words = {"int", "unsigned", "signed", "long", "short", "char", "_Bool", "bool"}
    for m in re.finditer(r"\b" + _DECL_TYPE + r"\s+([A-Za-z_]\w*)\s*(?=[=;,\[)])", body):
        name = m.group(1)
        if name in type_words or name in rename:
            continue
        rename[name] = f"{prefix}_l_{name}"


def _param_type(typ: str) -> str:
    t = " ".join((typ or "int").split())
    return t or "int"


def _build_inline_block(
    callee: FunctionInfo,
    args: list[str],
    *,
    prefix: str,
    tail: str,
) -> str | None:
    if len(args) != len(callee.params):
        return None
    decls: list[str] = []
    rename: dict[str, str] = {}
    for i, ((typ, pname), arg) in enumerate(zip(callee.params, args)):
        temp = f"{prefix}_i{i}"
        rename[pname] = temp
        decls.append(f"{_param_type(typ)} {temp} = {arg};")
    ret_var = f"{prefix}_ret" if _returns_value(callee) else None
    if ret_var:
        rt = _callee_ret_type(callee)
        if rt is None:
            return None
        decls.append(f"{rt} {ret_var};")
    _add_callee_locals(callee.body, prefix, rename)
    adapted = _adapt_callee_body(callee.body, rename, ret_var)
    if adapted is None:
        return None
    parts = decls + ([adapted] if adapted else [])
    if tail:
        parts.append(tail.rstrip(";") + ";")
    inner = "\n".join(parts)
    return "{\n" + inner + "\n}"


def _try_inline_stmt(
    stmt: str,
    fn: FunctionInfo,
    index: dict[tuple[str, str], FunctionInfo],
    site: int,
) -> str:
    raw = stmt
    s = stmt.strip()
    if not s.endswith(";"):
        return raw

    prefix = f"_h{site}"

    # return callee(...);
    m = re.match(r"return\s+", s)
    if m:
        call = _parse_call(s, m.end())
        if call is None:
            return raw
        name, args_src, _ = call
        if not s[m.end() :].lstrip().startswith(name):
            return raw
        callee = _lookup(index, fn.file, name)
        if callee is None or callee.name == fn.name or not _inlineable_callee(callee):
            return raw
        rest = s[call[2] :].strip()
        if rest != ";":
            return raw
        args = _split_args(args_src)
        block = _build_inline_block(
            callee,
            args,
            prefix=prefix,
            tail="return " + (f"{prefix}_ret" if _returns_value(callee) else "0"),
        )
        return block if block else raw

    # lhs = callee(...);
    m = re.match(r"([A-Za-z_]\w*)\s*=\s*", s)
    if m:
        lhs = m.group(1)
        call = _parse_call(s, m.end())
        if call is None:
            return raw
        name, args_src, end = call
        callee = _lookup(index, fn.file, name)
        if callee is None or callee.name == fn.name or not _inlineable_callee(callee):
            return raw
        if not _returns_value(callee):
            return raw
        rest = s[end:].strip()
        if rest != ";":
            return raw
        args = _split_args(args_src)
        block = _build_inline_block(
            callee,
            args,
            prefix=prefix,
            tail=f"{lhs} = {prefix}_ret",
        )
        return block if block else raw

    # type name = callee(...);
    m = re.match(
        rf"(?P<typ>{_SCALAR_TYPES})\s+(?P<name>[A-Za-z_]\w*)\s*=\s*",
        s,
    )
    if m:
        lhs = m.group("name")
        call = _parse_call(s, m.end())
        if call is None:
            return raw
        name, args_src, end = call
        callee = _lookup(index, fn.file, name)
        if callee is None or callee.name == fn.name or not _inlineable_callee(callee):
            return raw
        if not _returns_value(callee):
            return raw
        rest = s[end:].strip()
        if rest != ";":
            return raw
        args = _split_args(args_src)
        block = _build_inline_block(
            callee,
            args,
            prefix=prefix,
            tail=f"{lhs} = {prefix}_ret",
        )
        # The declaration stays in the caller's scope; the block assigns it.
        return f"{m.group('typ')} {lhs};\n{block}" if block else raw

    # callee(...);  void helper, no result use
    call = _parse_call(s, 0)
    if call is None:
        return raw
    name, args_src, end = call
    if s[end:].strip() != ";":
        return raw
    callee = _lookup(index, fn.file, name)
    if callee is None or callee.name == fn.name or not _inlineable_callee(callee):
        return raw
    if _returns_value(callee):
        return raw
    args = _split_args(args_src)
    block = _build_inline_block(callee, args, prefix=prefix, tail="")
    return block if block else raw


def _transform_body(
    body: str,
    fn: FunctionInfo,
    index: dict[tuple[str, str], FunctionInfo],
    counter: list[int],
) -> str:
    text = body
    out: list[str] = []
    while text:
        text = text.lstrip()
        if not text:
            break
        if text.startswith("{"):
            matched = _match_balanced(text, "{", "}")
            if matched is None:
                out.append(text)
                break
            inner, rest = matched
            out.append("{")
            out.append(_transform_body(inner, fn, index, counter))
            out.append("}")
            text = rest
            continue
        stmt, text = _take_stmt(text)
        if not stmt.strip():
            continue
        new_stmt = _try_inline_stmt(stmt, fn, index, counter[0])
        if new_stmt != stmt:
            counter[0] += 1
        out.append(new_stmt)
    return "".join(out)


def _inline_function(
    fn: FunctionInfo,
    index: dict[tuple[str, str], FunctionInfo],
) -> FunctionInfo:
    if fn.kind == "POINTER":
        return fn
    counter = [0]
    new_body = _transform_body(fn.body, fn, index, counter)
    if counter[0] == 0:
        return fn
    return replace(fn, body=new_body)


def inline_static(functions: list[FunctionInfo]) -> list[FunctionInfo]:
    """Inline static scalar callees (one level, same file) into caller bodies."""
    index = {(_basename(f.file), f.name): f for f in functions}
    return [_inline_function(fn, index) for fn in functions]
